plugins/transport/socket/main.go

```diff
@@ -115,50 +121,144 @@ func (s *Socket) initUDPSocket() *net.UDPConn {
-		for {
-			n, err := pc.Read(msgBuffer)
-			if err != nil || n < 1 {
-				if err != nil {
-					s.logger.Errorf(err, "reading from socket failed")
-				}
+
+	s.logger.Infof("socket listening on %s", s.conf.Socketaddr)
+
+	return pc
+}
+
+func (s *Socket) WriteTCPMsg(w transport.WriteFn, msgBuffer []byte, n int) (int64, error) {
+	var pos int64 = 0
+	var length int64
+	reader := bytes.NewReader(msgBuffer[:n])
+	for pos+msgLengthSize < int64(n) {
+		_, err := reader.Seek(pos, io.SeekStart)
+		if err != nil {
+			return pos, err
+		}
+		err = binary.Read(reader, binary.LittleEndian, &length)
+		if err != nil {
+			return pos, err
+		}
+
+		if pos+msgLengthSize+length > int64(n) ||
+			pos+msgLengthSize+length < 0 {
+			break
+		}
+		s.mutex.Lock()
+		w(msgBuffer[pos+msgLengthSize : pos+msgLengthSize+length])
+		msgCount++
+		s.mutex.Unlock()
+		pos += msgLengthSize + length
+	}
+	return pos, nil
+}
+
+func (s *Socket) ReceiveData(maxBuffSize int64, done chan bool, pc net.Conn, w transport.WriteFn) {
+	defer pc.Close()
+	msgBuffer := make([]byte, maxBuffSize)
+	var remainingMsg []byte
+	for {
+		n, err := pc.Read(msgBuffer)
+		if err != nil || n < 1 {
+			if err != nil {
+				s.logger.Errorf(err, "reading from socket failed")
+			}
+			if s.conf.Type != tcp {
 				done <- true
-				return
 			}
+			return
+		}
+		msgBuffer = append(remainingMsg, msgBuffer...)
 
-			// whole buffer was used, so we are potentially handling larger message
-			if n == len(msgBuffer) {
-				s.logger.Warnf("full read buffer used")
-			}
+		// whole buffer was used, so we are potentially handling larger message
+		if n == len(msgBuffer) {
+			s.logger.Warnf("full read buffer used")
+		}
 
-			if s.conf.DumpMessages.Enabled {
-				_, err := s.dumpBuf.Write(msgBuffer[:n])
-				if err != nil {
-					s.logger.Errorf(err, "writing to dump buffer")
-				}
-				_, err = s.dumpBuf.WriteString("\n")
-				if err != nil {
-					s.logger.Errorf(err, "writing to dump buffer")
-				}
-				s.dumpBuf.Flush()
+		n += len(remainingMsg)
+
+		if s.conf.DumpMessages.Enabled {
+			_, err := s.dumpBuf.Write(msgBuffer[:n])
+			if err != nil {
+				s.logger.Errorf(err, "writing to dump buffer")
+			}
+			_, err = s.dumpBuf.WriteString("\n")
+			if err != nil {
+				s.logger.Errorf(err, "writing to dump buffer")
 			}
+			s.dumpBuf.Flush()
+		}
 
+		if s.conf.Type == tcp {
+			parsed, err := s.WriteTCPMsg(w, msgBuffer, n)
+			if err != nil {
+				s.logger.Errorf(err, "error, while parsing messages")
+				return
+			}
+			remainingMsg = make([]byte, int64(n)-parsed)
+			copy(remainingMsg, msgBuffer[parsed:n])
+		} else {
 			w(msgBuffer[:n])
 			msgCount++
 		}
-	}(maxBufferSize)
+	}
+}
+
+// Run implements type Transport
+func (s *Socket) Run(ctx context.Context, w transport.WriteFn, done chan bool) {
+	var pc net.Conn
+	switch s.conf.Type {
+	case udp:
+		pc = s.initUDPSocket()
+		if pc == nil {
+			s.logger.Errorf(nil, "Failed to initialize socket transport plugin")
```

Comment:
```suggestion
			s.logger.Errorf(nil, "Failed to initialize UDP socket transport plugin")
```

Reply:
Perhaps this isn't even correct because this is the socket transport plugin, regardless of the connection method (`source`?). I might not be making a valid suggestion here, or maybe it should be phrased differently, but it feels like we should differentiate the connection type in the log message for future debugging.

Reply:
So what about something like: "Failed to initialize socket transport plugin with type: " and after that would be either TCP, UDP or unix?

Reply:
That works for me!